backend/main/models/vote.go

```diff
@@ -308,3 +328,145 @@ func DoesNFTExist(db *s.Database, v *VoteWithBalance) (bool, error) {
+			`
+			INSERT INTO community_users_achievements(addr, achievement_type, community_id, proposals, details)
+			VALUES($1, $2, $3, $4, $5)
+			ON CONFLICT (details)
+			DO NOTHING
+			RETURNING id
+		`, v.Addr, EarlyVote, p.Community_id, []int{v.Proposal_id}, earlyVoteDetails).Scan(&v.ID)
+
+		if checkErrorIgnoreNoRows(err) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func checkForStreakAchievement(db *s.Database, v *Vote, p Proposal) error {
+	var defaultStreakLength = 3
+
+	// Determine if this vote is part of a streak
+	// Proposals with the user address count as a vote for that proposal
+	// NULL means the user did not vote
+	sql := fmt.Sprintf(`
+		SELECT p.id as id, 
+		CASE WHEN v.addr is NULL THEN '' ELSE v.addr END as addr 
+		FROM proposals p 
+    LEFT OUTER JOIN (
+        SELECT * FROM votes where addr = '%s'
+    ) v ON v.proposal_id = p.id 
+    where p.community_id = $1 
+    ORDER BY start_time ASC
+		`, v.Addr)
+
+	var votingStreak []struct {
+		ID   uint64
+		Addr string
+	}
+	err := pgxscan.Select(db.Context, db.Conn, &votingStreak, sql, p.Community_id)
+
+	if checkError(err) {
+		return err
+	}
+
+	if len(votingStreak) >= defaultStreakLength {
+		var proposals []uint64
+		i := 0
+		for i < len(votingStreak) {
+			v := votingStreak[i]
+			if v.Addr != "" {
+				proposals = append(proposals, v.ID)
+			}
+			if len(proposals) >= defaultStreakLength && (i == len(votingStreak)-1 || (i < len(votingStreak)-1 && votingStreak[i+1].Addr == "")) {
+
+				//Unique identifier for current streak
+				currentStreakDetails := fmt.Sprintf("%s:%s:%d:%s", Streak, v.Addr, p.Community_id, strings.Trim(strings.Join(strings.Fields(fmt.Sprint(proposals)), ","), "[]"))
+
+				var sql string
+				if len(proposals) == defaultStreakLength {
+					// Add default streak, or do nothing if streak already exists
+					sql = `
+						INSERT INTO community_users_achievements(addr, achievement_type, community_id, proposals, details)
+						VALUES($1, $2, $3, $4, $5)
+						ON CONFLICT (details)
+						DO NOTHING
+					`
+					err = db.Conn.QueryRow(db.Context, sql, v.Addr, Streak, p.Community_id, proposals, currentStreakDetails).Scan(&v.ID)
+				} else {
+					// If previous streak exists, then update with new streak details
+					// e.g. If previous streak = 1,2,3 and current streak = 1,2,3,4 then update row with current streak details
+					previousStreakDetails := fmt.Sprintf("%s:%s:%d:%s", Streak, v.Addr, p.Community_id, strings.Trim(strings.Join(strings.Fields(fmt.Sprint(proposals[:len(proposals)-1])), ","), "[]"))
+					sql = `
+						INSERT INTO community_users_achievements(addr, achievement_type, community_id, proposals, details)
+						VALUES($1, $2, $3, $4, $5)
+						ON CONFLICT (details)
+						DO UPDATE SET proposals = $4, details = $6
+						RETURNING id
+					`
+					err = db.Conn.QueryRow(db.Context, sql, v.Addr, Streak, p.Community_id, proposals, previousStreakDetails, currentStreakDetails).Scan(&v.ID)
+				}
+
```

Comment:
I'm struggling to comprehend this block of code in the for loop. I think it's because of the amount of logic being exposed to the reader at one time. Maybe breaking it up will make it easier to read? So before you go into the for loop that could be done in a separate function with a meaningful name.

Also here
`// Add default streak, or do nothing if streak already exists`. Just include it in a function with a name like `AddDefaultStreak` then we might not need to add a comment, as the function name and parameters will be self-explanatory.

also ``// If previous streak exists, then update with new streak details`` new function call `updateStreakDetails` it will be easier to comprehend this way I think, and means each function is just doing one part of the logic. Will save other developers if they need to debug something.